### PR TITLE
Refactor IOState and PinData

### PR DESCRIPTION
**Pull request by martiniil**
## Description

The conversion of input/output bitset arrays to `PinState` vectors is postponed to calls of the user/ROS node (`IOState.input()`/`IOState.output()`). Internally `IOState` holds a `PinData` instance, which consists of two bitset arrays.

This is a preparation for #300.

ToDo:
- [x] Remove `IOState(std::vector<PinState>...)` constructor
- [x] `operator==`/`!=`, where is it really needed?
- [x] `ostream& operator<<`, where is it really needed?
- [x] Fix melodic build
- [x] Fix coverage

---

<details>
<summary>PR Checklist</summary>

### Things to add, update or check by the maintainers before this PR can be merged.

- [x] Public api function documentation
- [x] Architecture documentation reflects actual code structure
- [x] ~[Tutorials](https://wiki.ros.org/pilz_robots/Tutorials/)~
- [x] ~Overview on [ROS wiki](https://wiki.ros.org/pilz_robots)~
- [x] Package Readme ([example pilz_robots](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/README.md))
- [x] Good commit messages ([some tips](https://dev.to/jacobherrington/how-to-write-useful-commit-messages-my-commit-message-template-20n9))
- [x] CHANGELOG.rst updated
- [x] Copyright headers
- [x] ~Examples~

### Review Checklist
- [x] Soft- and hardware architecture (diagrams and description)
- [x] Test review (test plan and individual test cases)
- [x] Documentation describes purpose of file(s) and responsibilities
- [x] Code (coding rules, style guide)

### Release planning (please answer)
- [x] ~When is the new feature released?~
- [x] ~Which dependent packages do have to be released simultaneously?~

### Hardware tests
_Unstrike the text below to enable automatic hardware tests if available. Otherwise use this as a request for the reviewer._
 - [x] Perform hardware tests

</details>

### CODE REVIEW

File: standalone/test/include/psen_scan_v2_standalone/data_conversion_layer/io_pin_data_helper.h [hunk 1]
```diff
@@ -16,77 +16,43 @@
...
-  setPin(*it);
+  PinData pin_data{};
+  pin_data.inputState() = input_states;
+  pin_data.outputState() = output_states;
+  return pin_data;
 }
 }  // namespace psen_scan_v2_standalone_test
 
-#endif  // PSEN_SCAN_V2_STANDALONE_TEST_IO_PIN_DATA_HELPER_H
+#endif  // PSEN_SCAN_V2_STANDALONE_TEST_IO_PIN_DATA_HELPER_H
```
Comment (rfeistenauer) [inline standalone/test/include/psen_scan_v2_standalone/data_conversion_layer/io_pin_data_helper.h:58]: missing newline

File: standalone/test/unit_tests/api/unittest_io_state.cpp [hunk 2]
```diff
@@ -16,54 +16,68 @@
...
+    EXPECT_FALSE(output.state());
+  }
+}
 
-  EXPECT_EQ(io_state.output().size(), 1u);
-  EXPECT_EQ(io_state.output().at(0), PinState(5, "some name", false));
+TEST(IOStateTests, shouldReturnInputsEqualToConvertedInputPinData)
+{
+  PinData pin_data{};
+  pin_data.inputPinState(5, 1, true);  // make sure to not use "unused" bits
```
Comment (rfeistenauer) [inline standalone/test/unit_tests/api/unittest_io_state.cpp:53]: then maybe assert for it afterwards?

**Review by rfeistenauer — CHANGES_REQUESTED** [review]

File: test/unit_tests/unittest_io_state_rosconversions.cpp [hunk 3]
```diff
@@ -13,70 +13,132 @@
...
-  EXPECT_EQ(ros_message.output[1].state, false);
+TEST(IOStateRosConversionsTest, shouldAddAllOutputStatesOnce)
+{
+  psen_scan_v2_standalone::IOState io_state{};
+  psen_scan_v2::IOState ros_message = toIOStateMsg(io_state, "some_frame", 10 /* stamp */);
+
+  for (const auto& pin : io_state.output())
+  {
+    const auto pin_state_count{ countROSPinStatesWithId(ros_message.output, pin.id()) };
+    EXPECT_EQ(pin_state_count, 1) << "Wrong number of outputs with id " << pin.id() << " in the resulting Message";
```
Comment (martiniil) [inline test/unit_tests/unittest_io_state_rosconversions.cpp:102]:
```suggestion
    EXPECT_EQ(pin_state_count, 1u) << "Wrong number of outputs with id " << pin.id() << " in the resulting Message";
```

File: test/unit_tests/unittest_io_state_rosconversions.cpp [hunk 4]
```diff
@@ -13,70 +13,132 @@
...
-  EXPECT_EQ(ros_message.output[0].state, true);
+TEST(IOStateRosConversionsTest, shouldAddAllInputStatesOnce)
+{
+  psen_scan_v2_standalone::IOState io_state{};
+  psen_scan_v2::IOState ros_message = toIOStateMsg(io_state, "some_frame", 10 /* stamp */);
+
+  for (const auto& pin : io_state.input())
+  {
+    const auto pin_state_count{ countROSPinStatesWithId(ros_message.input, pin.id()) };
+    EXPECT_EQ(pin_state_count, 1) << "Wrong number of inputs with id " << pin.id() << " in the resulting Message";
```
Comment (martiniil) [inline test/unit_tests/unittest_io_state_rosconversions.cpp:90]:
```suggestion
    EXPECT_EQ(pin_state_count, 1u) << "Wrong number of inputs with id " << pin.id() << " in the resulting Message";
```

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/io_pin_data.h [hunk 5]
```diff
@@ -42,31 +39,115 @@ namespace monitoring_frame
...
+  std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES>& outputState();
+  const std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES>& inputState() const;
+  const std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES>& outputState() const;
+
+  bool operator==(const PinData& pin_data) const;
+
+private:
+  std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES> input_states_{};
+  std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES> output_states_{};
 };
```
Comment (martiniil) [inline standalone/include/psen_scan_v2_standalone/data_conversion_layer/io_pin_data.h:95]: We have now two ways of accessing the arrays with the same effect:
```
bool old_state = pin_data.inputPinState(4, 3);
pin_data.inputPinState(4, 3, true);
```
```
bool old_state = pin_data.inputState().at(4).test(3);
pin_data.inputState().at(4).set(3);
```
Do you think we have increased readability through the methods `inputPinState(...)`? If not, we could as well go back to the struct:
```
struct PinData
{
std::array<std::bitset<8>, NUMBER_OF_INPUT_BYTES> input_states{};
std::array<std::bitset<8>, NUMBER_OF_OUTPUT_BYTES> output_states{};
};
```

Comment (martiniil) [reply standalone/include/psen_scan_v2_standalone/data_conversion_layer/io_pin_data.h:95]: I think I like it better as a struct. The convenience methods are rarely used (in production code) anyways. The "set"-methods never and the "get"-methods only once.

**Review by rfeistenauer — APPROVED** [review]

**Pull request merged**